xdsl/dialects/onnx.py

```diff
@@ -26,6 +26,54 @@
 from xdsl.utils.exceptions import VerifyException
 
 
+def verify_shapes(
+    lhs_type: TensorType[Attribute] | list[int],
+    rhs_type: TensorType[Attribute] | list[int],
+    res_type: TensorType[Attribute],
+) -> None:
```

Comment:
I would recommend the following refactor:
```suggestion
def broadcast_shape(
    lhs: Sequence[int],
    rhs: Sequence[int],
) -> list[int] | None:
```
This will make it easier to test, it would be great to add some python unit tests for this helper in `tests/dialects/onnx/test_broadcast.py`

I would then recommend the following approach:
1. figure out which list is longer than the other
2. the prefix of the result is the prefix of the longer list
3. zip the shorter list with the suffix of the longer list
4. if the dimension is the same, then that's the dimension in the result, if one of them is 0, then take the other one, if they're not equal return `None`

The client of this function should extract the shapes from the types, check for `None`, and compare against the result type itself. Although maybe we may want a second helper to factor out that boilerplate too, but that helper itself would need no unit tests specifically.

Reply:
Helpfully, the docs from onnx provide a minimal set of test cases for the two broadcast shape calculation helpers

Reply:
i've attempted refactoring again, however here are my concerns:
1. i can't seem to figure out why onnx_invalid_mlir won't pass, the other one works fine
2. need help setting up pytest, i've checked the test cases in another editor and they all pass fine so thats not an issue, (i was getting the error that it cant recognise the tests)
3.  multidirectional broadcasting and unidirectional broadcasting seem to be the same thing to me? i ran the test cases for unidirectional broadcasting in the multidirectional broadcasting and the outputs work, so im not sure why there is a need for another computation.

I will paste my code here, if i am doing anything stupid or been missing something stupid, id appreciate the spot:

This method extracts the types
`def extract_shape_from_type(shape_type: [Union[Sequence[int], TensorType]]) -> Sequence[int] | None:
    if shape_type is None:
        return None
    elif isinstance(shape_type, TensorType):
        return shape_type.get_shape()
    else:
        return shape_type
`
This is the redone broadcast method (not sure if this is exacly how you wanted it done?):
`def broadcast_shape(lhs: Sequence[int], rhs: Sequence[int]) -> list[int] | None:
    """
    https://github.com/onnx/onnx/blob/main/docs/Broadcasting.md
    """
    lhs_shape = list(extract_shape_from_type(lhs))
    rhs_shape = list(extract_shape_from_type(rhs))

    # figure out which list is longer than the other
    if len(lhs_shape) > len(rhs_shape):
        longer_shape, shorter_shape = lhs_shape, rhs_shape
    else:
        longer_shape, shorter_shape = rhs_shape, lhs_shape
    # store the difference of shape dimensions between the shapes
    shape_dimension_diffs = len(longer_shape) - len(shorter_shape)
    # checking which side needs to be padded to the longer side
    if len(lhs_shape) != rhs_shape:
        shorter_shape = [1] * shape_dimension_diffs + list(shorter_shape)
    # Iterate over the shape dimensions and verify the broadcasting rules
    res_shape: list[int] = []
    for d1, d2 in zip(longer_shape, shorter_shape):
        if d1 == d2:
            res_shape.append(d1)
        elif d1 == 1:
            res_shape.append(d2)
        elif d2 == 1:
            res_shape.append(d1)
        else:
            raise VerifyException(
                f"operands have incompatible shapes: {shorter_shape} and {longer_shape}"
            )
    return res_shape`
(I was thinking to add the other check of result shape in the `verify_` of the opps,
All else i do is call the `broadcast_shape` method in class  method

I appreciate the feedback and help, @compor @superlopuh

Reply:
1. What happens when you run `make tests`?
2. I think I may have given you bad advice, or at least I got a better idea, it's probably best to copy the longer sequence into a new list, and then swap out the 1s with the other element or return None in a for loop.
3. It definitely seems to me like the two broadcast functions are different, Multidirectional seems symmetric, whereas Unidirectional is not, but it also seems like one can call the other after it determines which shape is shorter.